parquet-hadoop/src/main/java/org/apache/parquet/hadoop/CodecFactory.java

```diff
@@ -18,170 +18,83 @@
-      if (decompressor != null) {
-        CodecPool.returnDecompressor(decompressor);
-      }
-    }
+  public CodecFactory(Configuration configuration) {
+    this.configuration = configuration;
   }
 
-  /**
-   * Encapsulates the logic around hadoop compression
-   *
-   * @author Julien Le Dem
-   *
-   */
-  public static class BytesCompressor {
-
-    private final CompressionCodec codec;
-    private final Compressor compressor;
-    private final ByteArrayOutputStream compressedOutBuffer;
-    private final CompressionCodecName codecName;
-
-    public BytesCompressor(CompressionCodecName codecName, CompressionCodec codec, int pageSize) {
-      this.codecName = codecName;
-      this.codec = codec;
-      if (codec != null) {
-        this.compressor = CodecPool.getCompressor(codec);
-        this.compressedOutBuffer = new ByteArrayOutputStream(pageSize);
-      } else {
-        this.compressor = null;
-        this.compressedOutBuffer = null;
-      }
-    }
-
-    public BytesInput compress(BytesInput bytes) throws IOException {
-      final BytesInput compressedBytes;
-      if (codec == null) {
-        compressedBytes = bytes;
-      } else {
-        compressedOutBuffer.reset();
-        if (compressor != null) {
-          // null compressor for non-native gzip
-          compressor.reset();
-        }
-        CompressionOutputStream cos = codec.createOutputStream(compressedOutBuffer, compressor);
-        bytes.writeAllTo(cos);
-        cos.finish();
-        cos.close();
-        compressedBytes = BytesInput.from(compressedOutBuffer);
-      }
-      return compressedBytes;
-    }
+  public C getCompressor(CompressionCodecName codecName, int pageSize) {
+    C comp = compressors.get(codecName);
+    if (comp == null) {
 
-    private void release() {
-      if (compressor != null) {
-        CodecPool.returnCompressor(compressor);
-      }
+      CompressionCodec codec = getCodec(codecName);
+      comp = createCompressor(codecName, codec, pageSize);
+      compressors.put(codecName, comp);
     }
+    return comp;
+  }
 
-    public CompressionCodecName getCodecName() {
-      return codecName;
+  public D getDecompressor(CompressionCodecName codecName) {
+    D decomp = decompressors.get(codecName);
+    if (decomp == null) {
+      CompressionCodec codec = getCodec(codecName);
+      decomp = createDecompressor(codec);
+      decompressors.put(codecName, decomp);
     }
-
+    return decomp;
   }
 
-  private final Map<CompressionCodecName, BytesCompressor> compressors = new HashMap<CompressionCodecName, BytesCompressor>();
-  private final Map<CompressionCodecName, BytesDecompressor> decompressors = new HashMap<CompressionCodecName, BytesDecompressor>();
-  private final Map<String, CompressionCodec> codecByName = new HashMap<String, CompressionCodec>();
-  private final Configuration configuration;
+  protected C createCompressor(CompressionCodecName codecName, CompressionCodec codec, int pageSize) {return null;}
```

Comment:
I ended up moving the implementations of both createDecompressor and createCompressor up to this class and removing the HeapCodecFactory. If the codec name is passed into these, the lookup of the codecs themselves needs to happen in a non-static context to have access to the configuration set in the constructor of the factory. The HeapBytescompressor is used by the DirectCodecFactory for fallback, so it needs to be able to create one there. I can break them back out if you'd prefer, but it was one solution to the concerns and it moves these classes back to where they were originally.